feat(api): add ramp rate to protocol engine's execute profile

ryanthecoder opened the pull request:
# Overview
 The other way to control the thermocycler's block temperature besides set_block_temperature is to use execute_profile, which takes in a list of ThermocyclerStep
 
 This PR adds ramp rate as an optional to each ThermocyclerStep so users can use it if they want.

ryanthecoder committed:
fixup some of the validation.

ryanthecoder committed:
add validation decoys to tests

ryanthecoder committed:
format

sfoster1 commented on shared-data/command/schemas/14.json:
```diff
@@ -4162,9 +4162,14 @@
           "description": "Time to hold target temperature in seconds.",
           "title": "Holdseconds",
           "type": "number"
+        },
```
should be in schema 15 not 14

sfoster1 commented on api/src/opentrons/protocol_engine/commands/thermocycler/run_extended_profile.py:
```diff
@@ -33,6 +33,9 @@ class ProfileStep(BaseModel):
     holdSeconds: float = Field(
         ..., description="Time to hold target temperature in seconds."
     )
+    rampRate: float = Field(
```
i think these need a default or we won't be able to take old protocol command records out of the database. try it on a robot I guess

sfoster1 replied:
even if so, can we make these use "None" as unlimited and turn it into a 0-is-unlimited thing lower down since it's essentially about how we encode it more than anything else?

sfoster1 commented on api/src/opentrons/protocol_engine/commands/thermocycler/run_profile.py:
```diff
@@ -30,6 +30,9 @@ class RunProfileStepParams(BaseModel):
         ..., description="Time to hold target temperature at in seconds."
     )
+    rampRate: float = Field(
+        ..., description="How quickly to change temperature in °C/second."
```
night need a default, see above

sfoster1 commented on api/src/opentrons/protocol_api/validation.py:
```diff
@@ -489,17 +489,24 @@ def ensure_thermocycler_profile_steps(
                 "defined for each step in cycle"
             )
+        if ramp_rate is None:
+            validated_ramp_rate = 0.0
```
is 0 really the default we want? do we have special handling somewhere to make that mean "unlimited"?

ryanthecoder replied:
0.0 is what the firmware uses to signify infinite. I'll redefine it in python so it's not magic

sfoster1 reviewed (changes requested):
Couple little things I think we can improve, and I'm concerned about how we signify unlimited rates

ryanthecoder committed:
swtich to a none-means-inifinit scheme.

ryanthecoder committed:
Setup ramprate through the protocol engines execute profile method of thermocycler control.

ryanthecoder committed:
update schema

sfoster1 commented on api/src/opentrons/protocol_api/validation.py:
```diff
@@ -489,17 +489,22 @@ def ensure_thermocycler_profile_steps(
                 "defined for each step in cycle"
             )
+        if ramp_rate is not None and ramp_rate <= 0:
+            raise ValueError("Ramp rate must be greater than or equal to 0.")
```
should it allow 0? 0 seems degenerate

ryanthecoder replied:
had the right logic and wrong comment

sfoster1 commented on api/src/opentrons/protocol_engine/state/module_substates/thermocycler_module_substate.py:
```diff
@@ -170,7 +175,9 @@ def validate_ramp_rate(self, ramp_rate: float, target_temp: float) -> float:
         if ramp_rate <= 0:
-            raise InvalidRampRateError("Thermocycler ramp rate cannot be negative or 0")
+            raise InvalidRampRateError(
+                f"Thermocycler ramp rate cannot be less than or equalt to 0, got {ramp_rate}"
```
```suggestion
                f"Thermocycler ramp rate cannot be less than or equal to 0, got {ramp_rate}"
```
Probably shouldn't allow 0 here though.

ryanthecoder replied:
this does not allow 0

sfoster1 reviewed (changes requested):
imo we should not allow 0 here - it would lead to an infinite profile right?

ryanthecoder committed:
format schema

ryanthecoder committed:
fix comments around validation.

sfoster1 reviewed (approved):
Okay, nice! lgtm!

ryanthecoder committed:
Merge branch 'edge' into EXEC-1833-ramp-rate-in-tc-profile

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal EXEC-1833-ramp-rate-in-tc-profile snapshots (#19537)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/19494

Co-authored-by: ryanthecoder <5885321+ryanthecoder@users.noreply.github.com>

ryanthecoder committed:
add defaults to profile runs.

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal EXEC-1833-ramp-rate-in-tc-profile snapshots (#19543)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/19494

Co-authored-by: ryanthecoder <5885321+ryanthecoder@users.noreply.github.com>

ryanthecoder committed:
command schema
